File: scripts/R/prep02_smallarea_crosswalk.R
# -*- coding: utf-8 -*-
# 소지역 중심좌표(X_COORD/Y_COORD) → 섬코드 매핑 (R 버전).
#   소지역 테이블에 좌표가 있으므로 별도 집계구 경계 shp 불필요.
#   점(블록 중심) → island_boundaries 폴리곤 point-in-polygon 공간조인.
# 산출: 04_center/smallarea_island_lookup.csv  (카드·SKT 병합키)
#       04_center/adong_island_map.csv         (행정동→섬, 다대다)
source("C:/Users/dyu18/OneDrive/문서/데이터활용대회/scripts/R/prep00_config.R")

NEAREST_FALLBACK_M <- 0   # >0 이면 섬 경계 밖 블록을 그 거리 내 최근접 섬에 보강

sa <- smart_read_csv(FILE_SMALLAREA) %>%
  filter(SIDO_CD == SIDO_JEONNAM) %>%
  mutate(X_COORD = as_num(X_COORD), Y_COORD = as_num(Y_COORD),
         AREA = as_num(AREA)) %>%
  filter(!is.na(X_COORD), !is.na(Y_COORD))

pts <- st_as_sf(sa, coords = c("X_COORD", "Y_COORD"), crs = SMALLAREA_CRS) %>%
  st_transform(ISLAND_CRS)
isl <- st_read(file.path(INTERIM, "island_boundaries_jeonnam.gpkg"), quiet = TRUE) %>%
  st_transform(ISLAND_CRS) %>% select(섬코드, 섬이름)

joined <- st_join(pts, isl, join = st_within, left = TRUE)

# (옵션) 섬 경계 밖 블록 보강
if (NEAREST_FALLBACK_M > 0) {
  miss <- joined %>% filter(is.na(섬코드))
  if (nrow(miss) > 0) {
    nidx <- st_nearest_feature(miss, isl)
    d <- as.numeric(st_distance(miss, isl[nidx, ], by_element = TRUE))
    ok <- d <= NEAREST_FALLBACK_M
    joined$섬코드[is.na(joined$섬코드)][ok] <- isl$섬코드[nidx][ok]
    joined$섬이름[is.na(joined$섬이름)][ok] <- isl$섬이름[nidx][ok]
  }
}

lookup <- joined %>% st_drop_geometry()
keep <- intersect(c("BLOCK_CD","SMRY_AREA","ADONG_CD","ADONG_NM",
                    "LDONG_CD","SGNG_CD","SGNG_NM","AREA","섬코드","섬이름"),
                  names(lookup))
lookup <- lookup[, keep]
write_excel_csv(lookup, file.path(OUT, "smallarea_island_lookup.csv"))

cat("=== 소지역 → 섬 크로스워크 ===\n")
cat(sprintf("전남 블록 %d | 섬 매칭 %d | 커버 섬 %d\n",
            nrow(lookup), sum(!is.na(lookup$섬코드)),
            length(unique(na.omit(lookup$섬코드)))))

# 행정동 → 섬 (다대다)
adong <- lookup %>% filter(!is.na(섬코드)) %>%
  group_by(ADONG_CD) %>%
  summarise(섬코드들 = paste(sort(unique(as.character(섬코드))), collapse = ","),
            .groups = "drop")
write_excel_csv(adong, file.path(OUT, "adong_island_map.csv"))
cat("행정동→섬 매핑 저장:", nrow(adong), "개\n저장:",
    file.path(OUT, "smallarea_island_lookup.csv"), "\n")

File: scripts/Python/prep02_smallarea_crosswalk.py
# -*- coding: utf-8 -*-
"""
소지역 중심좌표(X_COORD/Y_COORD) → 섬코드 매핑.
  핵심: 소지역 테이블에 좌표가 있으므로 별도 집계구 경계 shp가 필요 없다.
  점(블록 중심) → 기존 island_boundaries 폴리곤에 point-in-polygon 공간조인.

산출:
  04_center/smallarea_island_lookup.csv  (BLOCK_CD/집계구/행정동 ↔ 섬코드)  ★카드·SKT 병합키
  04_center/adong_island_map.csv         (행정동 → 섬코드 리스트, 다대다·안분/맥락용)
"""
import geopandas as gpd
import pandas as pd

from prep00_config import (FILE_SMALLAREA, INTERIM, OUT, ISLAND_CRS,
                           SIDO_JEONNAM, SMALLAREA_CRS, smart_read_csv)

NEAREST_FALLBACK_M = 0   # >0 으로 두면 섬 경계 밖 블록을 그 거리 내 최근접 섬에 보강매칭

sa = smart_read_csv(FILE_SMALLAREA)
sa = sa[sa["SIDO_CD"] == SIDO_JEONNAM].copy()
for c in ["X_COORD", "Y_COORD", "AREA", "LENGTH"]:
    if c in sa.columns:
        sa[c] = pd.to_numeric(sa[c], errors="coerce")

pts = gpd.GeoDataFrame(
    sa, geometry=gpd.points_from_xy(sa["X_COORD"], sa["Y_COORD"]),
    crs=SMALLAREA_CRS).to_crs(ISLAND_CRS)
isl = gpd.read_file(INTERIM / "island_boundaries_jeonnam.gpkg").to_crs(ISLAND_CRS)

joined = gpd.sjoin(pts, isl[["섬코드", "섬이름", "geometry"]],
                   how="left", predicate="within").drop(columns="index_right")

# (옵션) 섬 경계 밖 블록을 최근접 섬에 보강 — 부두/방파제 등 경계 살짝 밖 블록 구제
if NEAREST_FALLBACK_M > 0:
    miss = joined[joined["섬코드"].isna()].copy()
    if len(miss):
        near = gpd.sjoin_nearest(
            miss.drop(columns=["섬코드", "섬이름"]),
            isl[["섬코드", "섬이름", "geometry"]],
            how="left", max_distance=NEAREST_FALLBACK_M, distance_col="_d")
        joined.loc[near.index, "섬코드"] = near["섬코드"]
        joined.loc[near.index, "섬이름"] = near["섬이름"]

keep = [c for c in ["BLOCK_CD", "SMRY_AREA", "ADONG_CD", "ADONG_NM",
                    "LDONG_CD", "SGNG_CD", "SGNG_NM", "AREA", "섬코드", "섬이름"]
        if c in joined.columns]
lookup = joined[keep].copy()
lookup.to_csv(OUT / "smallarea_island_lookup.csv", index=False, encoding="utf-8-sig")

print("=== 소지역 → 섬 크로스워크 ===")
print(f"전남 블록 {len(lookup)} | 섬 매칭 {lookup['섬코드'].notna().sum()} "
      f"| 커버 섬 {lookup['섬코드'].nunique()}")

# 행정동 → 섬(다대다). 한 면이 여러 섬을 포함하므로 안분/맥락용으로 별도 보관.
adong = (lookup.dropna(subset=["섬코드"])
         .groupby("ADONG_CD")["섬코드"]
         .agg(lambda s: ",".join(sorted(set(s.astype(str))))).reset_index())
adong.to_csv(OUT / "adong_island_map.csv", index=False, encoding="utf-8-sig")
print("행정동→섬 매핑 저장:", len(adong), "개 행정동")
print("저장:", OUT / "smallarea_island_lookup.csv")
